### PR TITLE
Add IM subscribe request/response builder and parser

### LINKED ISSUE #7977: Add SubscribeRequest/Response Builder/Parser support
Following the spec https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#subscriberequestmessage to implement builder and parser for subscribe request and response

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
What is being fixed?  Examples:
* Per IM encoding spec, add subscribe request/response builder/parser(https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#subscriberequestmessage )
* Fixes #7977

#### Change overview
Add IM subscribe request/response builder and parser
Add ReturnLogErrorOnFailure and VerifyOrReturnLogError to log the error in original place.
#### Testing
Add unit test to cover subscriber request/response builder and parser

### CODE REVIEW

File: src/app/MessageDef/SubscribeRequest.h [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+     *  @param [in] aEventNumber The event number
+     *  @return A reference to *this
+     */
+    SubscribeRequest::Builder & EventNumber(const uint64_t aEventNumber);
+
+    /**
+     *  @brief The requested minimum interval boundary floor in seconds
+     *  @return A reference to *this
+     */
+    SubscribeRequest::Builder & MinInterval(const uint16_t aMinInterval);
```
Comment (andy31415) [inline src/app/MessageDef/SubscribeRequest.h:183]: Rename to `MinIntervalSeconds` and then doccomment is not needed anymore and usage will be clearer everywhere. Same for maxinterval.

File: src/app/MessageDef/SubscribeRequest.h [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+     *  @brief Initialize a AttributeDataVersionList::Builder for writing into the TLV stream
+     *
+     *  @return A reference to AttributeDataVersionList::Builder
+     */
+    AttributeDataVersionList::Builder & CreateAttributeDataVersionListBuilder();
+
+    /**
+     *  @brief An initiator can optionally specify an EventNumber it has already to limit the
+     *  set of retrieved events on the server for optimization purposes.
+     *  @param [in] aEventNumber The event number
```
Comment (andy31415) [inline src/app/MessageDef/SubscribeRequest.h:174]: remove `@param` item - it is obvious that EventNumber/aEventNumber refers to an event number.

File: src/app/MessageDef/SubscribeRequest.h [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+     *  @return A reference to *this
+     */
+    SubscribeRequest::Builder & EndOfSubscribeRequest();
+
+private:
+    AttributePathList::Builder mAttributePathListBuilder;
+    EventPathList::Builder mEventPathListBuilder;
+    AttributeDataVersionList::Builder mAttributeDataVersionListBuilder;
+};
+}; // namespace SubscribeRequest
```
Comment (andy31415) [inline src/app/MessageDef/SubscribeRequest.h:218]: remove redundant `;`

File: src/app/tests/TestMessageDef.cpp [hunk 4]
```diff
@@ -28,6 +28,8 @@
 #include <app/MessageDef/InvokeCommand.h>
 #include <app/MessageDef/ReadRequest.h>
 #include <app/MessageDef/ReportData.h>
+#include <app/MessageDef/SubscribeRequest.h>
```
Comment (andy31415) [inline src/app/tests/TestMessageDef.cpp:31]: should we consider splitting the tests into separate classes? We have the messagedef already split however the test is still a monolith

File: src/app/MessageDef/SubscribeRequest.cpp [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+    // make a copy of the reader here
+    mReader.Init(aReader);
+
+    VerifyOrExit(chip::TLV::kTLVType_Structure == mReader.GetType(), err = CHIP_ERROR_WRONG_TLV_TYPE);
+
+    err = mReader.EnterContainer(mOuterContainerType);
+
+exit:
+    ChipLogFunctError(err);
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.cpp]: Agree, just introduce VerifyOrReturnLogError and ReturnLogErrorOnFailure which can log error in original failure points

File: src/app/MessageDef/SubscribeRequest.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::TLV::TLVReader reader;
+
+    err = mReader.FindElementWithTag(chip::TLV::ContextTag(kCsTag_AttributePathList), reader);
+    SuccessOrExit(err);
+
+    VerifyOrExit(chip::TLV::kTLVType_Array == reader.GetType(), err = CHIP_ERROR_WRONG_TLV_TYPE);
+
+    err = apAttributePathList->Init(reader);
+    SuccessOrExit(err);
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.cpp]: removed, thanks

File: src/app/MessageDef/SubscribeRequest.cpp [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+
+CHIP_ERROR SubscribeRequest::Builder::Init(chip::TLV::TLVWriter * const apWriter)
+{
+    return InitAnonymousStructure(apWriter);
+}
+
+AttributePathList::Builder & SubscribeRequest::Builder::CreateAttributePathListBuilder()
+{
+    // skip if error has already been set
+    VerifyOrExit(CHIP_NO_ERROR == mError, mAttributePathListBuilder.ResetError(mError));
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.cpp]: https://github.com/project-chip/connectedhomeip/issues/8219, would handle this cleanup in MessageDef in separate ticket, thanks

File: src/app/MessageDef/SubscribeRequest.cpp [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+    }
+    return *this;
+}
+
+SubscribeRequest::Builder & SubscribeRequest::Builder::EndOfSubscribeRequest()
+{
+    EndOfContainer();
+    return *this;
+}
+}; // namespace app
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.cpp]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.h [hunk 9]
```diff
@@ -0,0 +1,220 @@
+/**
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: removed, thanks

File: src/app/MessageDef/SubscribeRequest.h [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    kCsTag_MaxInterval               = 5,
+    kCsTag_KeepExistingSubscriptions = 6,
+    kCsTag_IsProxy                   = 7,
+};
+
+class Parser : public chip::app::Parser
+{
+public:
+    /**
+     *  @brief Initialize the parser object with TLVReader
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.h [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+     *          #CHIP_END_OF_TLV if there is no such element
+     */
+    CHIP_ERROR GetIsProxy(bool * const apIsProxy) const;
+};
+
+class Builder : public chip::app::Builder
+{
+public:
+    /**
+     *  @brief Initialize a SubscribeRequest::Builder for writing into a TLV stream
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.h [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+     *   1) all mandatory tags are present
+     *   2) all elements have expected data type
+     *   3) any tag can only appear once
+     *   4) At the top level of the structure, unknown tags are ignored for forward compatibility
+     *  @note The main use of this function is to print out what we're
+     *    receiving during protocol development and debugging.
+     *    The encoding rule has changed in IM encoding spec so this
+     *    check is only "roughly" conformant now.
+     *
+     *  @return #CHIP_NO_ERROR on success
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.h (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.h (hunk 3 above)
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h:181]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.h (hunk 2 above)
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: Done, thanks

File: src/app/tests/TestMessageDef.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/tests/TestMessageDef.cpp:31]: https://github.com/project-chip/connectedhomeip/issues/8220
Agree, would handle it in separate ticket, thanks

File: src/app/MessageDef/SubscribeRequest.h [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+     *  @param [in] apWriter    A pointer to TLVWriter
+     *
+     *  @return #CHIP_NO_ERROR on success
+     */
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+
+    /**
+     *  @brief Initialize a AttributePathList::Builder for writing into the TLV stream
+     *
+     *  @return A reference to AttributePathList::Builder
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.h]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.cpp [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+{
+    // skip if error has already been set
+    if (mError == CHIP_NO_ERROR)
+    {
+        mError = mAttributePathListBuilder.Init(mpWriter, kCsTag_AttributePathList);
+        ChipLogFunctError(mError);
+    }
+    else
+    {
+        mAttributePathListBuilder.ResetError(mError);
```
Comment (andy31415) [inline src/app/MessageDef/SubscribeRequest.cpp:233]: would it be the same and simpler to follow if we always reset?

```cpp
mAttributePathListBuilder.ResetError(mError);

if (mError == CHIP_NO_ERROR) {
    mError = mAttributePathListBuilder.Init(mpWriter, kCsTag_AttributePathList);
    ChipLogFunctError(mError);
}

return mAttributePathListBuilder;
```

or is this logically different (i.e. the path builder may contain an error and we do not want to clear it with the Request error)?

I find the logic of containing several separate errors a bit hard to follow. What is the invariant that we try to maintain regarding the relationship between mError and mAttributePathListBuilder::mError? What do combinations mean of "both no error", "both error", "one error, one no error"? Are the 2 variants of "one has error, one does not have error" even valid?

Comment (andy31415) [reply src/app/MessageDef/SubscribeRequest.cpp:233]: Since this pattern is used throughout, we have errors for:

mError (data member)
mAttributePathListBuilder::mError
mEventPathListBuilder::mError
mAttributeDataVersionListBuilder::mError

so the combinations of having/not having errors seems to expand exponentially.
Still wondering what the invariant is - it seems to me we would want a single 'global error' instead of 4 separate ones. Is this correct? Could we desing our code to have such an approach?

File: src/app/MessageDef/SubscribeResponse.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+     *  @brief Final Sync Interval for the subscription back to the clients.
+     */
+    SubscribeResponse::Builder & FinalSyncIntervalSeconds(const uint16_t aFinalSyncIntervalSeconds);
+
+    /**
+     *  @brief Mark the end of this SubscribeResponse
+     */
+    SubscribeResponse::Builder & EndOfSubscribeResponse();
+};
+}; // namespace SubscribeResponse
```
Comment (andy31415) [inline src/app/MessageDef/SubscribeResponse.h:97]: nit: remove extra  ; for namespace ends

Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeResponse.h]: Done, thanks

File: src/app/MessageDef/SubscribeRequest.cpp (hunk 14 above)
Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.cpp]: just remove mAttributePathListBuilder.ResetError(mError), there is no need to pass the error inside nested container
currently they are separete, and maintained for particular container, for example,
AttributeDataVersionList::Builder attributeDataVersionList = subscribeRequestBuilder.CreateAttributeDataVersionListBuilder();
NL_TEST_ASSERT(apSuite, subscribeRequestBuilder.GetError() == CHIP_NO_ERROR); //  the rror is inside top container
    aAttributeDataVersionListBuilder.AddVersion(1).aAttributeDataVersionListBuilder.EndOfAttributeDataVersionList();
    NL_TEST_ASSERT(apSuite, aAttributeDataVersionListBuilder.GetError() == CHIP_NO_ERROR);
===> the error is processed inside nested container.

I am still thinking of how to optimize the nested layer for better pattern

Comment (yunhanw-google) [reply src/app/MessageDef/SubscribeRequest.cpp]: https://github.com/project-chip/connectedhomeip/issues/8228

**Review by jmartinez-silabs — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]
